npe2/_inspection/_visitors.py

```diff
@@ -0,0 +1,455 @@
+import ast
+import inspect
+from abc import ABC, abstractmethod
+from importlib.metadata import Distribution
+from pathlib import Path
+from types import ModuleType
+from typing import TYPE_CHECKING, Any, DefaultDict, Dict, List, Tuple, Type, Union
+
+from ..manifest import contributions
+
+if TYPE_CHECKING:
+    from pydantic import BaseModel
+
+_COMMAND_PARAMS = inspect.signature(contributions.CommandContribution).parameters
+
+CONTRIB_MAP: Dict[str, Tuple[Type["BaseModel"], str]] = {
+    "writer": (contributions.WriterContribution, "writers"),
+    "reader": (contributions.ReaderContribution, "readers"),
+    "sample_data_generator": (contributions.SampleDataGenerator, "sample_data"),
+    "widget": (contributions.WidgetContribution, "widgets"),
+}
+
+
+class _DecoratorVisitor(ast.NodeVisitor, ABC):
```

Comment:
this class just split out from NPE2PluginModuleVisitor from `main` ...  the only change made is the `match` parameter that lets you specify what name to match